Implement recursion and 2d-pose videos

julien-h opened the pull request:
This fixes the `--unlabeled-recursive` flag and implements pose-2d videos.

**[1]** To run in recursive mode:

```
export FOLDER=/ramdya-nas/CLC/190904_SS51017-tdTomGC6fopt/Fly3/CO2xzGG/

df3d-cli $FOLDER --unlabeled-recursive
```

I have also created a small test template in `./data`:

```
cd /home/semihg/workspace/DeepFly3D
source activate deepfly
df3d-cli ./data/recursive --unlabeled-recursive
```

**[2]** To limit the number of images to `100`:

```
df3d-cli $FOLDER --unlabeled-recursive -n 100
```

**[3]** The script will look for the subfolders named `images`, for instance:

```
$FOLDER/behData_001/images/
$FOLDER/behData_002/images/
```
And will create a subfolder `images/df3d/` containing the predictions and the video.

**[4]** To view all the generated videos in the NAS, you can use this command:

```
cd /ramdya-nas/CLC/190904_SS51017-tdTomGC6fopt/Fly3/CO2xzGG/
vlc `find . -name "*.mp4" | tr "\n" " "`
```

**[5]**  I launched the predictions and video generation recursively in `$FOLDER` Friday evening at 20:00. Hopefully the script will have completed without errors during the weekend and you can check the videos on monday. For extra clarity, here's the command I ran:

```
df3d-cli /ramdya-nas/CLC/190904_SS51017-tdTomGC6fopt/Fly3/CO2xzGG/ --unlabeled-recursive
```

If using the session `semihg`, you can connect to the running `tmux` using:

```
tmux a -t 1
```

**Remarks**
Fixing the path to `df3d` and the `--recursive-flag` was more involved than I would have thought. Mainly because in some part of the code the variable `output_folder` contains only the relative path `output_folder = df3d/` supposed to be relative to the `input_folder`, while in others it is set to the full path: `output_folder = path.join(input_folder, 'df3d')`.

semihgunel commented:
1. please move the camera re-ordering inside the loop.
2. can you replace def get_max_img_id(path) with get_max_img_id from os_util? it is under https://github.com/NeLy-EPFL/DeepFly3D/blob/master/deepfly/GUI/util/os_util.py

julien-h commented:
I don't understand for the `get_max_img_id` thing, that's already the function I'm using:

```
from ..GUI.util.os_util import get_max_img_id
```

julien-h commented:
I have changed the code to use regex though. Do you want the old code back? I don't think a binary search makes it faster than acessing the os once to get the directory listing then filtering with a pre-compiled regex

semihgunel commented:
I am getting an error:
![image](https://user-images.githubusercontent.com/20509861/67211045-63ab2b80-f41a-11e9-9baa-801dd44c2d2d.png)

semihgunel commented:
> I have changed the code to use regex though. Do you want the old code back? I don't think a binary search makes it faster than acessing the os once to get the directory listing then filtering with a pre-compiled regex

It makes it much faster as getting full directory list is very slow (~10 seconds) on large directories.

julien-h commented:
(1) I have reverted the code to use the binary search.
(2) The error in your screenshot id probably because you don't use the right virtual environment. In the debug output, it searches for images in the `images/df3d` folders, which is a bug from the old version of `find_leaves_recursive` that I have changed

semihgunel commented:
Currently, the saved videos are too large (2gb). You can resize them t something like small (around 500x500) to reduce size.

julien-h commented:
**This commit implements:**
1. resizing for 2d videos
2. 3d videos

**Two things left to do** about the 3D plots:
1. rotate the 3d plot because it's currently upside down
2. the procrustes thing

I have looked into the procrustes code, but it's still unclear with what data I should call the method.

**The code relevant for the 3d plots is:**
- `make_pose3d_video(args) --> imgs_generator()` [link](https://github.com/julien-h/DeepFly3D/blob/front/deepfly/CLI/main.py#L186)
- `compute_3d_img` [link](https://github.com/julien-h/DeepFly3D/blob/front/deepfly/CLI/main.py#L273)


**About the generator**

I have found a way to reuse the same code to create videos for both 2d and 3d. To do that I created an inner method `img_generator()` which is a generator. 

A generator is simply a method that will create a sequence on the fly, using the keyword `yield`. We can iterate over this sequence like we would with any other.

semihgunel commented:
Great work!

julien-h commented:
This last commit is Friday's work, fixing the 3d plots calibration/procrustes/smoothness.
I ran the code recursively on the NAS using these commands:

```
$ cd /ramdya-nas/CLC/190904_SS51017-tdTomGC6fopt/Fly3/CO2xzGG
$ df3d-cli . --unlabeled-recursive --vid3d
```

Let's see on monday.
